worker/codalabworker/worker.py

```diff
@@ -242,13 +230,22 @@ def remove_dependency(self, parent_uuid, parent_path, uuid):
         self._dependency_manager.remove_dependency(parent_uuid, parent_path, uuid)
 
     def _read(self, socket_id, uuid, path, read_args):
+        socket = self._bundle_service.socket(worker_id=self.id, socket_id=socket_id)
         run = self._worker_state_manager._get_run(uuid)
         if run is None:
-            Run.read_run_missing(self._bundle_service, self, socket_id)
+            message = {
+                'error_code': httplib.INTERNAL_SERVER_ERROR,
+                'error_message': BUNDLE_NO_LONGER_RUNNING_MESSAGE,
+            }
+            socket.reply(message)
+        elif run.read(path=path, read_args=read_args, socket=socket):
+            pass
         else:
-            # Reads may take a long time, so do the read in a separate thread.
-            threading.Thread(target=Run.read,
-                             args=(run, socket_id, path, read_args)).start()
+            message = {
+                'error_code': httplib.INTERNAL_SERVER_ERROR,
+                'error_message': 'Read failed for unknown reason.',
+            }
+            socket.reply(message)
```

Comment:
It seems a little weird to send something to the socket after you've already passed it to someone else to send data to, since they could presumably fail after sending some data (e.g. possibly an error message). Can we just remove the return value from run.read (since it's always True right now anyway) and say it's up to the implementer of run.read to send either some data or an error message?

Reply:
I agree. I'm just going to remove this and document this in `read`